# HW 2 Part 1: Association Rule Mining on Bike Rentals

The original dataset is available here: https://s3.amazonaws.com/baywheels-data/index.html 
We have a filtered subset of this dataset for this homework. 

The dataset is preprocessed to include all the locations from which rides on a day and hour started in 2021. If you completed HW 1 part 3, you can extract the same from your own code. The notebook below also contains code to preprocess is further for association rule mining. 

Perform the following tasks on the provided bike rental dataset. Complete all the [LP] questions to receive a "low pass" grade on the homework. The [HP] question specifications combine all parts of the homework. Please refer to Canvas for more details.

Note that if you are unable to complete any of the LP questions satisfactorily, you will receive a grade of "revision required". You can revise and resubmit your work in exchange for a token. Please review the syllabus for more information on specifications grading.

**VERY IMPORTANT**: Include **ALL** the references you used for this assignment, including names of classmates you discuss with. Failure to cite your sources counts as an act of academic dishonesty and will be taken seriously without zero tolerance. You will automatically receive a “fail” grade in the homework and further serious penalties may be imposed.

NOTE: You can look for help on the Internet but refrain from referencing too much. Please cite all your sources in your submission. 
When you submit your assignment, you automatically agree to the following statement. If you do not agree, it is your responsibility to provide the reason.

“*I affirm that I have neither given nor received unauthorized help in completing this homework. I am not aware of others receiving such help. I have cited all the sources in the solution file.*”

In [1]:
import pandas as pd
## Replace path with your file's path
path = "../Part1/baywheels_ride_locations_2021.csv"
stations_imported = pd.read_csv(path, index_col=0, converters={"start_station_list": lambda x: x.strip("[]").replace("'","").replace("\"","").split(", ")})
stations_imported.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8748 entries, 0 to 8747
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   start_date          8748 non-null   object
 1   start_hour          8748 non-null   int64 
 2   start_station_list  8748 non-null   object
dtypes: int64(1), object(2)
memory usage: 273.4+ KB


In [2]:
s = stations_imported['start_station_list'].explode()
b = stations_imported[['start_date', 'start_hour']].join(pd.crosstab(s.index, s))
basket = b.drop(columns=['start_date', 'start_hour'])
for col in basket:
  basket[col] = basket[col].astype('bool')

In [ ]:
## Your code goes here. Import the csv into a pandas dataframe here

In [3]:
basket.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8748 entries, 0 to 8747
Columns: 482 entries, 10th Ave at E 15th St to Yerba Buena Center for the Arts (Howard St at 3rd St)
dtypes: bool(482)
memory usage: 4.3 MB


In [4]:
basket.head(25)

10th Ave at E 15th St  10th Ave at Irving St  10th St at Empire St  \
0                   False                  False                 False   
1                   False                  False                 False   
2                   False                  False                 False   
3                   False                  False                 False   
4                   False                  False                 False   
5                   False                  False                 False   
6                   False                   True                 False   
7                   False                  False                 False   
8                   False                  False                 False   
9                   False                  False                 False   
10                  False                   True                 False   
11                  False                   True                 False   
12                  False                   True                 False   
13                  False                   True                 False   
14                  False                   True                 False   
15                  False                   True                 False   
16                  False                   True                 False   
17                  False                   True                 False   
18                  False                  False                 False   
19                  False                  False                  True   
20                  False                  False                 False   
21                  False                  False                 False   
22                  False                  False                 False   
23                  False                  False                 False   
24                  False                  False                 False   

    10th St at Fallon St  10th St at Mission St  10th St at University Ave  \
0                  False                  False                      False   
1                  False                  False                      False   
2                  False                  False                      False   
3                  False                  False                      False   
4                  False                  False                      False   
5                  False                  False                      False   
6                  False                  False                      False   
7                  False                  False                      False   
8                  False                  False                      False   
9                  False                  False                      False   
10                 False                  False                      False   
11                 False                  False                      False   
12                 False                  False                      False   
13                 False                  False                       True   
14                  True                  False                      False   
15                 False                  False                      False   
16                 False                  False                      False   
17                 False                  False                      False   
18                 False                  False                      False   
19                 False                  False                      False   
20                 False                  False                      False   
21                 False                  False                      False   
22                 False                  False                      False   
23                 False                  False                      False   
24                 False                  False                      False   

    11th St at Bryant St  11th St at Natoma St  13

[LP] Perform Association Rule Mining using the apriori algorithm on the `basket` dataframe with a support above 0.28 and confidence above 0.9.

The library you can use is mlxtend. The imports are included below. Answer the questions listed below after you perform the apriori algorithm. Answer ALL questions correctly to pass this question.

1.   Q1. The number of frequent itemsets of length 1: 
2.   Q2. The number of frequent itemsets of length 4: 
3.   Q3. The number of frequent itemsets of length 5: 
4.   Q4. How many association rules meet the confidence requirement:





### Answer to LP

1.   The number of frequent itemsets of length 1 is 142. 
2.   The number of frequent itemsets of length 4 is 421. 
3.   The number of frequent itemsets of length 5 is 1. 
4.   The number of association rules that meet the confidence requirement is 1157. 

In [5]:
## Your code for LP goes here. 
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(basket, min_support=0.28, use_colnames=True)

frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support                                           itemsets  length
0     0.309442                            (10th Ave at Irving St)       1
1     0.356996                             (11th St at Bryant St)       1
2     0.370713                             (11th St at Natoma St)       1
3     0.308871                            (14th St at Mission St)       1
4     0.301440                   (16th St Mission BART Station 2)       1
...        ...                                                ...     ...
3719  0.290009  (Powell St BART Station (Market St at 5th St),...       4
3720  0.287380  (Powell St BART Station (Market St at 5th St),...       4
3721  0.285094  (Powell St BART Station (Market St at 5th St),...       4
3722  0.285780  (Powell St BART Station (Market St at 5th St),...       4
3723  0.281093  (Powell St BART Station (Market St at 5th St),...       5

[3724 rows x 3 columns]

In [6]:
frequent_itemsets[(frequent_itemsets['length'] == 1) & (frequent_itemsets['support'] >= 0.28)]

support                                           itemsets  length
0    0.309442                            (10th Ave at Irving St)       1
1    0.356996                             (11th St at Bryant St)       1
2    0.370713                             (11th St at Natoma St)       1
3    0.308871                            (14th St at Mission St)       1
4    0.301440                   (16th St Mission BART Station 2)       1
..        ...                                                ...     ...
137  0.393919                       (Washington St at Kearny St)       1
138  0.318244                            (Webster St at Clay St)       1
139  0.295839                           (Webster St at Grove St)       1
140  0.280407                        (Webster St at OFarrell St)       1
141  0.325332  (Yerba Buena Center for the Arts (Howard St at...       1

[142 rows x 3 columns]

In [7]:
frequent_itemsets[(frequent_itemsets['length'] == 2) & (frequent_itemsets['support'] >= 0.28)]

support                                           itemsets  length
142   0.303612       (11th St at Bryant St, Market St at 10th St)       2
143   0.281436  (11th St at Bryant St, Powell St BART Station ...       2
144   0.316987       (Market St at 10th St, 11th St at Natoma St)       2
145   0.286008  (11th St at Natoma St, Powell St BART Station ...       2
146   0.282007  (Powell St BART Station (Market St at 5th St),...       2
...        ...                                                ...     ...
1530  0.284065   (Valencia St at 22nd St, Valencia St at 16th St)       2
1531  0.298011  (Valencia St at Clinton Park, Valencia St at 1...       2
1532  0.280636   (Valencia St at 22nd St, Valencia St at 21st St)       2
1533  0.297554  (Valencia St at Clinton Park, Valencia St at 2...       2
1534  0.284294  (Valencia St at 22nd St, Valencia St at Clinto...       2

[1393 rows x 3 columns]

In [8]:
frequent_itemsets[(frequent_itemsets['length'] == 3) & (frequent_itemsets['support'] >= 0.28)]

support                                           itemsets  length
1535  0.282579  (Market St at 10th St, 16th St at Prosper St, ...       3
1536  0.291267  (18th St at Noe St, Market St at 10th St, 17th...       3
1537  0.283150  (1st St at Folsom St, 17th St at Valencia St, ...       3
1538  0.289666  (1st St at Folsom St, Berry St at 4th St, 17th...       3
1539  0.287723  (1st St at Folsom St, 17th St at Valencia St, ...       3
...        ...                                                ...     ...
3297  0.291152  (Powell St BART Station (Market St at 5th St),...       3
3298  0.284522  (Powell St BART Station (Market St at 5th St),...       3
3299  0.284751  (Powell St BART Station (Market St at 5th St),...       3
3300  0.289666  (Powell St BART Station (Market St at 5th St),...       3
3301  0.297096  (San Francisco Caltrain (Townsend St at 4th St...       3

[1767 rows x 3 columns]

In [18]:
frequent_itemsets[(frequent_itemsets['length'] == 4) & (frequent_itemsets['support'] >= 0.28)]

support                                           itemsets  length
3302  0.280293  (1st St at Folsom St, Market St at Dolores St,...       4
3303  0.296754  (1st St at Folsom St, Market St at 10th St, 17...       4
3304  0.291610  (Powell St BART Station (Market St at 5th St),...       4
3305  0.298925  (17th St at Valencia St, Market St at 10th St,...       4
3306  0.294010  (Powell St BART Station (Market St at 5th St),...       4
...        ...                                                ...     ...
3718  0.290352  (Powell St BART Station (Market St at 5th St),...       4
3719  0.290009  (Powell St BART Station (Market St at 5th St),...       4
3720  0.287380  (Powell St BART Station (Market St at 5th St),...       4
3721  0.285094  (Powell St BART Station (Market St at 5th St),...       4
3722  0.285780  (Powell St BART Station (Market St at 5th St),...       4

[421 rows x 3 columns]

In [10]:
f5 = frequent_itemsets[(frequent_itemsets['length'] == 5) & (frequent_itemsets['support'] >= 0.28)]
f5

support                                           itemsets  length
3723  0.281093  (Powell St BART Station (Market St at 5th St),...       5

In [20]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)

rules

antecedents  \
0                            (22nd St Caltrain Station)   
1                              (7th Ave at Cabrillo St)   
2                                (8th St at Brannan St)   
3                        (Alameda St at Henry Adams St)   
4                                (Brannan St at 7th St)   
...                                                 ...   
1152  (Union Square (Powell St at Post St), Sanchez ...   
1153  (Powell St BART Station (Market St at 5th St),...   
1154  (Powell St BART Station (Market St at 5th St),...   
1155  (Powell St BART Station (Market St at 5th St),...   
1156  (Powell St BART Station (Market St at 5th St),...   

                 consequents  antecedent support  consequent support  \
0     (Market St at 10th St)            0.376315             0.66701   
1     (Market St at 10th St)            0.362940             0.66701   
2     (Market St at 10th St)            0.376772             0.66701   
3     (Market St at 10th St)            0.341907             0.66701   
4     (Market St at 10th St)            0.379858             0.66701   
...                      ...                 ...                 ...   
1152  (Market St at 10th St)            0.331619             0.66701   
1153  (Market St at 10th St)            0.312643             0.66701   
1154  (Market St at 10th St)            0.309214             0.66701   
1155  (Market St at 10th St)            0.328418             0.66701   
1156  (Market St at 10th St)            0.295725             0.66701   

       support  confidence      lift  leverage  conviction  
0     0.340764    0.905529  1.357594  0.089758    3.524773  
1     0.326703    0.900157  1.349542  0.084619    3.335156  
2     0.339849    0.902002  1.352308  0.088539    3.397945  
3     0.308413    0.902039  1.352364  0.080358    3.399230  
4     0.344193    0.906109  1.358465  0.090824    3.546561  
...        ...         ...       ...       ...         ...  
1152  0.311157    0.938297  1.406722  0.089964    5.396677  
1153  0.294010    0.940402  1.409878  0.085474    5.587293  
1154  0.291038    0.941220  1.411104  0.084790    5.665025  
1155  0.307499    0.936304  1.403733  0.088441    5.227767  
1156  0.281093    0.950522  1.425050  0.083842    6.730048  

[1157 rows x 9 columns]

[HP] Explain briefly what the frequent itemsets and association rules mean. Your answer needs to be specific for this dataset. You do not need to mention the definition of these terms. Mention how these insights can be used - e.g., by the bike rental service provider or the city.

In [ ]:
## Your code for HP goes here. 

In [87]:
toplocations = frequent_itemsets.sort_values('support' , ascending = False)[:10]
toplocations


support                                           itemsets  length
89   0.667010                             (Market St at 10th St)       1
122  0.586648                            (Sanchez St at 15th St)       1
110  0.582419     (Powell St BART Station (Market St at 5th St))       1
90   0.577046                          (Market St at Dolores St)       1
109  0.575446     (Powell St BART Station (Market St at 4th St))       1
8    0.574417                           (17th St at Valencia St)       1
129  0.540352              (Union Square (Powell St at Post St))       1
48   0.523548  (Civic Center/UN Plaza BART Station (Market St...       1
98   0.516575                               (Mission Playground)       1
34   0.512346                               (Berry St at 4th St)       1

In [83]:
f5_locations = f5.itemsets
list(f5_locations)

[frozenset({'17th St at Valencia St',
            'Market St at 10th St',
            'Market St at Dolores St',
            'Powell St BART Station (Market St at 4th St)',
            'Powell St BART Station (Market St at 5th St)'})]

In [69]:
top_rules = rules.sort_values('confidence', ascending=False)[:10]
top_rules

antecedents  \
876   (Powell St BART Station (Market St at 5th St),...   
890   (4th St at Mission Bay Blvd S, Berry St at 4th...   
942   (Esprit Park, Berry St at 4th St, Powell St BA...   
946   (Powell St BART Station (Market St at 5th St),...   
1043  (Jackson Playground, Market St at Steuart St, ...   
891   (Berry St at 4th St, 4th St at Mission Bay Blv...   
958   (Powell St BART Station (Market St at 5th St),...   
957   (Berry St at 4th St, Market St at Steuart St, ...   
776   (Berry St at 4th St, 17th St at Valencia St, M...   
875   (4th St at 16th St, Berry St at 4th St, Powell...   

                 consequents  antecedent support  consequent support  \
876   (Market St at 10th St)            0.298354             0.66701   
890   (Market St at 10th St)            0.301326             0.66701   
942   (Market St at 10th St)            0.295153             0.66701   
946   (Market St at 10th St)            0.294582             0.66701   
1043  (Market St at 10th St)            0.294010             0.66701   
891   (Market St at 10th St)            0.323274             0.66701   
958   (Market St at 10th St)            0.318701             0.66701   
957   (Market St at 10th St)            0.332876             0.66701   
776   (Market St at 10th St)            0.316301             0.66701   
875   (Market St at 10th St)            0.308528             0.66701   

       support  confidence      lift  leverage  conviction  
876   0.285322    0.956322  1.433745  0.086317    7.623728  
890   0.288066    0.955994  1.433254  0.087078    7.566920  
942   0.281664    0.954299  1.430713  0.084794    7.286281  
946   0.280979    0.953822  1.429998  0.084490    7.211061  
1043  0.280407    0.953733  1.429863  0.084299    7.197070  
891   0.308185    0.953324  1.429251  0.092558    7.134067  
958   0.303727    0.953013  1.428784  0.091150    7.086849  
957   0.316987    0.952266  1.427665  0.094955    6.976029  
776   0.301097    0.951934  1.427166  0.090122    6.927702  
875   0.293667    0.951834  1.427017  0.087876    6.913393

* For frequent itemsets of length 1, we have a total of 142 frequent itemsets that meet the minimum support. 
* For frequent itemsets of length 4, we have a total of 421 number of frequent itemsets that meet the minimum support. 
* For frequent itemsets of length 5, we have only one frequent itemset that meets the minimum support. 

* Using the frequent itemsets, we can generate our association rules, and a total of 1157 rules were generated.

Looking at the consequent in all rules, "Market St at 10th St" with the highest support of 66.7%, also leads to a higher confidence of 95.6%, which seems to be a favorable spot that the company could promote, add more assessts to, and/or keep as is for the bay wheels to retain their customers and gain new customers. "Powell St BART Station" at any Market St and Berry St at 4th St also seem to be locations popular amongst the customers which implies that customers would most likely also pick Market St at 10th St as their location. These two locations might be driving majority of the sales for the bay wheels company. The company can try to boost their sales or return of investments if they continue to improve their customer experience by providing improved service in these locations. 



# References
1. http://rasbt.github.io/mlxtend/api_subpackages/mlxtend.frequent_patterns/
2.  http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/#example-1-generating-frequent-itemsets
3. http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/#example-1-generating-association-rules-from-frequent-itemsets

# What to turn in:
1. The ipynb solution file, which includes the references
2. If you use Colab or GitHub for version control, please share a link to your notebook or GitHub repository
https://github.com/ankitajaswal/Cmpe255/tree/main/HW2


